docs/defaulting.md

```diff
@@ -0,0 +1,249 @@
+    </controller>
+    <controller type='pci' index='0' model='pci-root'/>
+    <interface type='network'>
+      <mac address='52:54:00:2e:84:30'/>
+      <source network='default'/>
+      <model type='rtl8139'/>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x02' function='0x0'/>
+    </interface>
+    <input type='mouse' bus='ps2'/>
+    <input type='keyboard' bus='ps2'/>
+    <memballoon model='virtio'>
+      <address type='pci' domain='0x0000' bus='0x00' slot='0x03' function='0x0'/>
+    </memballoon>
+  </devices>
+</domain>
+```
+
+Finally, when a Domain was successfully started by libvirt, libvirt assigns an
+`alias` to every device in the XML.
+
+## The Problem
+
+ * Hotplug operations are difficult with the minimal VM spec of kubevirt,
+   since we could match wrong devices.
+ * After VM restarts, different devices or addresses could be assigned to a VM
+   by libvirt
+ * libvirt assigns all that data on the host, either when defining the Domain,
+   or after the Domain was started. That requires asynchronous lookups and in
+   the worst case even merge logic.
+ * If we don't want to change Guest OS relevant fields in the VM specification
+   behind the back of the user, all computed default values by libvirt, need to
+   be reflected in the VM status.
+
+If your VMs are ephemeral (no migrations, recreating VMs vs. restarting VMs, no
+hotplug), you are not affected by that.
+
+## Proposed Solution
+
+### Defaulting service based on libvirt
+
+Except for some runtime information, which is normally not included in
+migratable domain XMLs, libvirt fills in all defaults when defining a domain,
+which influence how the emulated machine looks like for the Guest OS. We can
+leverage that with a simple REST based service, which defines VMs in an
+isolated libvirt (running libvirt and the small REST service in a Pod). The
+service is a standalone component in KubeVirt and can be integrated in the
+following flow:
+
+ * When doing a POST to
+   `/apis/kubevirt.io/v1alpha1/namespaces/mynamespace/vms`, `virt-api` can do
+   a roundtrip to that service and prefill the defaults. As the response of
+   the POST to `virt-api`, you will get the completely prefilled VM
+   specification. This updated spec contains all the necessary bits, to allow
+   consistent hotplugging and keeping the VM definition consistent between
+   redefines.
+
+Additional flows, like a dedicated `dry-run` endpoint, to allow management
+applications, reusing existing admission controller logic for prefilling VMs,
+withouth having to actually run a VM are thinkable. That would allow complete
+decoupling of managment and runtime.
+
+Defaulting rules:
+
+ * The Kubevirt VM spec can contain abstract definitions, like a reference to a
+   volume claim, or a reference to a host network. In such cases most of the
+   time, the device `source` section needs to be replaced by a dummy `source`
```

Comment:
Like with Pods, I would only aim for backend configuration abstractions. The rest can and should IMHO be covered by admission controllers before the defaulter is invoked.